Moving Remove Tensor Type Specializations to after custom passes

Gamrix opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71748

This is to allow for Intel folks to use type information in their custom passes.

Differential Revision: [D35513184](https://our.internmc.facebook.com/intern/diff/D35513184)

Gamrix committed:
Moving Remove Tensor Type Specializations to after custom passes

This is to allow for Intel folks to use type information in their custom passes.

[ghstack-poisoned]

eellison reviewed (commented):
I don't think this is enough - we also run `removeTensorTypeSpecializations(graph_->block());` in TE Fuser so I think this persists the problem. Can you add a test ? (maybe with custom post pass ?)

eellison commented:
Bump on this - @chunyuan-w and other intel devs were requesting

chunyuan-w commented:
cc @jgong5

Gamrix committed:
Update on "Moving Remove Tensor Type Specializations to after custom passes"


This is to allow for Intel folks to use type information in their custom passes.

[ghstack-poisoned]

Gamrix committed:
Update on "Moving Remove Tensor Type Specializations to after custom passes"


This is to allow for Intel folks to use type information in their custom passes.

[ghstack-poisoned]

Gamrix committed:
Update on "Moving Remove Tensor Type Specializations to after custom passes"


This is to allow for Intel folks to use type information in their custom passes.

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -141,6 +141,67 @@ bool isSupported(Node* node) {
 }
 
+// Code to detect specializations for testing CustomPasses.
+static std::unique_ptr<RegisterPass> tTSpecializationDetectionPass = nullptr;
```
Code for testing shouldn't leak into our actual implementation files, especially when we have an open registration mechanism.

Gamrix replied:
Which file do you think the code should go into instead then?

eellison replied:
register a custom pass in a test file

eellison commented on torch/csrc/jit/runtime/graph_executor.cpp:
```diff
@@ -915,6 +915,10 @@ void runNondiffOptimization(
   for (const auto& passPair : getCustomPostPasses()) {
     passPair.first(graph);
   }
+  if(tensorExprFuserEnabled()){
```
This is in the legacy executor, we don't need to run `RemoveTensorTypeSpecializations` here because the types are guaranteed by the inputs, since we cache and specialize. I've been trying to delete it but unfortunately there's a legacy glow integration internally :/

eellison commented on torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:
```diff
@@ -453,15 +453,8 @@ void ProfilingGraphExecutorImpl::runNoGradOptimizations(
-      GRAPH_DEBUG(
-          "After RemoveTensorTypeSpecializations, before customPostPasses\n",
-          *graph);
+      FuseTensorExprs(graph, getFusionGroupInlining() ? 2 : 1);
```
We should still be calling `FuseTensorExprs(graph, min_size, /*composed_op*/ false, dyn_shapes);`

You removed the `dyn_shapes` input here

eellison commented on torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:
```diff
@@ -475,9 +468,14 @@ void ProfilingGraphExecutorImpl::runNoGradOptimizations(
       passPair.first(graph);
     }
+    GRAPH_DEBUG("After customPostPasses \n", *graph);
+    if (tensorExprFuserEnabled()) {
```
We want to call `RemoveTensorTypeSpecializations` regardless of whether or not `TE` is enabled. Let's make it the contract that the individaul `customPostPasses` don't need to invoke it themselves, that way they will compose. 

Should we also remove it here ? cc @jjsjann123  https://github.com/pytorch/pytorch/blob/0dc3f829d9d11ca7e972cb5fcc95059742fa71ea/torch/csrc/jit/codegen/cuda/graph_fuser.cpp#L2042

jjsjann123 replied:
Yes we can also remove the pass used in nvfuser.

eellison commented on torch/csrc/jit/runtime/static/fusion.cpp:
```diff
@@ -347,6 +347,7 @@ void performTensorExprFusion(
       /*min_group_size*/ 2,
       /*add_composed_op*/ false,
       /*fuse_to_dynamic_shapes*/ true);
+  RemoveTensorTypeSpecializations(graph);
```
Nice!

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -141,6 +141,67 @@ bool isSupported(Node* node) {
+  return false;
+}
+
+void detectTTSpecializationPass(std::shared_ptr<Graph>& graph) {
```
All of this code should be registered as a custom pass in a C++ test file, instead of having the test leak into our implementation

eellison reviewed (commented):
looks good, a few commetns.

Gamrix committed:
Update on "Moving Remove Tensor Type Specializations to after custom passes"


This is to allow for Intel folks to use type information in their custom passes.

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/utils/check_tensor_specializations.cpp:
```diff
@@ -0,0 +1,67 @@
+#include <torch/csrc/jit/passes/utils/check_tensor_specializations.h>
+#include <torch/csrc/jit/jit_log.h>
+
+namespace torch {
```
This is still leaking into source code instead of test

Gamrix replied:
As per our discussion, I changed the test to a C-based test so that we can have all of the testing code in test files.

eellison reviewed (commented):
This is almost the same thing as the previous PR. If you go to the [blame of pass_manager.cpp](https://github.com/pytorch/pytorch/blame/master/torch/csrc/jit/passes/pass_manager.cpp) and then click through a couple links you'll see the [original PR](https://github.com/pytorch/pytorch/pull/18587/files) that added this. 

As is custom, the [code to test it](https://github.com/pytorch/pytorch/blob/cfd92f2d59f97bf90529d1e8c37ba0467ce1c566/test/cpp/jit/test_misc.cpp#L1530) was added in a test file instead of source code.

Gamrix committed:
Update on "Moving Remove Tensor Type Specializations to after custom passes"


This is to allow for Intel folks to use type information in their custom passes.

[ghstack-poisoned]

Gamrix committed:
Update on "Moving Remove Tensor Type Specializations to after custom passes"


This is to allow for Intel folks to use type information in their custom passes.

[ghstack-poisoned]

Gamrix committed:
Update on "Moving Remove Tensor Type Specializations to after custom passes"


This is to allow for Intel folks to use type information in their custom passes.

[ghstack-poisoned]

eellison reviewed (approved):
Sweet! LGTM !!

Gamrix commented:
@pytorchbot merge this please

eellison commented:
PR like this we should probably be checking about internal tests

eellison commented:
at least import it

Gamrix committed:
Update on "Moving Remove Tensor Type Specializations to after custom passes"


This is to allow for Intel folks to use type information in their custom passes.

[ghstack-poisoned]

Gamrix committed:
Update on "Moving Remove Tensor Type Specializations to after custom passes"


This is to allow for Intel folks to use type information in their custom passes.

[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35513184).
